[AutoDiff] Optimize derivative generic signature computation.

dan-zheng opened the pull request:
Computing constrained derivative generic signatures is costly: remove
unnecessary uses in the differentiation transform. The only remaining use
is `[differentiable]` attribute construction, where it is necessary for
correctness. Other users now extract the constrained derivative generic
signature from a `[differentiable]` attribute.

---

This should improve compiler performance. Will report benchmark results before merging.

dan-zheng committed:
[AutoDiff] Optimize constrained derivative generic signature computation.

Computing constrained derivative generic signatures is costly: remove
unnecessary uses in the differentiation transform. The only remaining use
is `[differentiable]` attribute construction, where it is necessary for
correctness. Other users now extract the constrained derivative generic
signature from a `[differentiable]` attribute.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -269,7 +264,19 @@ static CanGenericSignature getConstrainedDerivativeGenericSignature(
+      nullptr);
+}
+
+// Returns the canonical derivative generic signature for the given
```
1. Three slashes for doc comment. 
2. “Returns” -> “Return” for consistency.

dan-zheng replied:
> 1. Three slashes for doc comment.

Thanks for the catch, fixed.

> 2. “Returns” -> “Return” for consistency.

"Returns" is actually used more frequently than "Return" in this file (26 vs 4). I would prefer "Returns" in line with Apple Swift documentation ([example](https://developer.apple.com/documentation/swift/array/2994747-randomelement)). What would you prefer?

rxwei replied:
Most functions in this file are not using a third person plural. However, I’m +1 on using third person plurals.

dan-zheng committed:
Address review feedback.

Use triple slashes for doc comment.

dan-zheng commented:
`tensorflow/swift-apis` Kokoro CI found that this patch crashes fast.ai Swift notebooks:
```
swift: /swift-base/swift/include/swift/AST/Requirement.h:87: swift::Type swift::Requirement::getSecondType() const: Assertion `getKind() != RequirementKind::Layout' failed.
Stack dump:
0.	Program arguments: /swift-tensorflow-toolchain/usr/bin/swift -frontend -c /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/00_load_data.swift /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/01_matmul.swift /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/01a_fastai_layers.swift /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/02_fully_connected.swift /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/02a_why_sqrt5.swift /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/03_minibatch_training.swift -primary-file /fastai_dev/swift/FastaiNotebook_04_callbacks/Sources/FastaiNotebook_04_callbacks/04_callbacks.swift -emit-module-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/FastaiNotebook_04_callbacks.build/04_callbacks~partial.swiftmodule -emit-module-doc-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/FastaiNotebook_04_callbacks.build/04_callbacks~partial.swiftdoc -emit-dependencies-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/FastaiNotebook_04_callbacks.build/04_callbacks.d -emit-reference-dependencies-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/FastaiNotebook_04_callbacks.build/04_callbacks.swiftdeps -target x86_64-unknown-linux -disable-objc-interop -I /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug -enable-testing -g -module-cache-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/ModuleCache -swift-version 4.2 -Onone -D SWIFT_PACKAGE -D DEBUG -enable-anonymous-context-mangled-names -parse-as-library -module-name FastaiNotebook_04_callbacks -o /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/FastaiNotebook_04_callbacks.build/04_callbacks.swift.o -index-store-path /fastai_dev/swift/FastaiNotebook_11_imagenette/.build/x86_64-unknown-linux/debug/index/store -index-system-modules 
1.	Swift version 5.1-dev (LLVM c5340df2d1, Swift 0301b1c4f4)
2.	While running pass #1956 SILModuleTransform "Differentiation".
```

I'll attempt to reproduce. If the failure is reproducible, I'll fix and add a regression test.

dan-zheng committed:
Check layout/superclass constraints in derivative generic signatures.

Derivative generic signatures may now contain layout/superclass constraints.
This occurs for `autodiff_function` instructions with original function
operands that are generic with layout/superclass constraints.

Add tests.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow macOS

dan-zheng commented:
The variance in Kokoro compilation benchmark results seems too high to be useful, unfortunately. I'll manually benchmark performance locally.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2265,27 +2275,56 @@ static bool diagnoseUnsatisfiedRequirements(ADContext &context,
+      }
     }
     switch (req.getKind()) {
+    // Check layout requirements.
```
Note: I couldn't think of ways to test unsatisfied layout requirements. This should be easier after [TF-864](https://bugs.swift.org/browse/TF-864), when `@differentiable` attribute supports layout requirements in where clauses.
